RPI4 led name changed from led0 to ACT?

eriknordmark opened the pull request:
I don't know what and when this changed, but led0 no longer exists.
Instead we have PWR and ACT.

rvs reviewed (changes requested):
The only reason this could change is via DeviceTree update. This brings me to my 2nd point -- since we don't fully control DT for RPi4 I suggest you actually add a block instead of replacing it so we don't ruin those units that may still be working fine with older DTs.

eriknordmark commented:
> The only reason this could change is via DeviceTree update. This brings me to my 2nd point -- since we don't fully control DT for RPi4 I suggest you actually add a block instead of replacing it so we don't ruin those units that may still be working fine with older DTs.

But wouldn't a change in the DT come with a new version of EVE?
It isn't like I changed the DT intentionally.

And since this is a raspberrypi.rpi.raspberrypi,4-model-b.brcm,bcm2711 (and not a different model), if we want to support multiple options we'd have to list multiple LEDs and have ledmanager try the listed LEDs until it finds one which exists. Is that a reasonable approach?

giggsoff commented:
I can see some recent changes of dts inside raspberry/linux fork. https://github.com/raspberrypi/linux/commit/22201d4868f66aa7773d5c4bad7bab1795ee6757#diff-2401e76a6c909c36337ef376d13ec770c079dc46ee9394cc52dce7eacd2d36cf

eriknordmark commented:
> I can see some recent changes of dts inside raspberry/linux fork. [raspberrypi/linux@22201d4#diff-2401e76a6c909c36337ef376d13ec770c079dc46ee9394cc52dce7eacd2d36cf](https://github.com/raspberrypi/linux/commit/22201d4868f66aa7773d5c4bad7bab1795ee6757#diff-2401e76a6c909c36337ef376d13ec770c079dc46ee9394cc52dce7eacd2d36cf)

But that doesn't have a label of "ACT" for "led0". I have to look at what we have now in EVE as DTS.

rvs commented:
Hey @eriknordmark -- is this still relevant or can we close this?

rene commented:
The reason for the ACT label it's because we get the device tree file for RPi from [u-boot](https://github.com/lf-edge/eve/blob/master/pkg/u-boot/Dockerfile#L64), and following the includes in the device tree, we reach to the following line:

https://github.com/u-boot/u-boot/blob/v2022.10/arch/arm/dts/bcm2835-rpi.dtsi#L8

That defines the label ACT. So the LED must be renamed on EVE. However, I agree that must be backward compatible (for the reasons already pointed out here).

rene commented on pkg/pillar/cmd/ledmanager/ledmanager.go:
```diff
@@ -194,13 +194,13 @@ var mToF = []modelToFuncs{
 		initFunc:    InitLedCmd,
 		displayFunc: ExecuteLedCmd,
-		arg:         "led0",
+		arg:         "ACT",
```
I think this commit should be done after the other commit that introduces the possibility of list of different names for the same LED...

rene commented on pkg/pillar/cmd/ledmanager/ledmanager.go:
```diff
@@ -194,13 +196,15 @@ var mToF = []modelToFuncs{
 		displayFunc: ExecuteLedCmd,
-		arg:         "ACT",
+		// Try led0 first and if not found, ACT
+		arg: "led0,ACT",
```
This commit is introducing a nice feature, which is the support of different possible names for the same LED. I won't change any LED name here. I think it's better to do it in the other commit, this one should focus only in the new feature. Also, ACT is the default now, so I would change the other: "ACT,led0".

rene commented on pkg/pillar/cmd/ledmanager/ledmanager.go:
```diff
@@ -633,14 +638,24 @@ const (
+	for _, led := range leds {
+		triggerFilename := fmt.Sprintf("/sys/class/leds/%s/trigger", led)
+		b := []byte("none")
+		err := ioutil.WriteFile(triggerFilename, b, 0644)
```
Ideally we should fail if the trigger file doesn't exist. However, I think there is no problem here to change file mode to 0600 to make yetus happy since in real-world the file should be always there (created by led platform driver).

PS: ioutil is deprecated by I think Milan's PR already took care of this.

rene commented on pkg/pillar/cmd/ledmanager/ledmanager.go:
```diff
@@ -754,6 +769,10 @@ func doLedAction(ledName string, turnon bool) {
 // Enable the led for 200ms
 func doLedBlink(ledName string) {
+	if ledName == "" {
+		time.Sleep(200 * time.Millisecond)
```
I got the point here. We don't want to mess up time aware functions if some LED doesn't exist, right? Well, but at the other hand it can mask the problem in case some LED doesn't exist to the callers. Maybe would be better to explicitly put this delay in the callers in case of error, instead let "hidden" here....

eriknordmark commented:
@rene  it's been a while and i dont have my rpi with me. Can you check that this pr still fixes it?

rene commented:
> @rene it's been a while and i dont have my rpi with me. Can you check that this pr still fixes it?

@eriknordmark , sure, we have 3 RPis in the office, I will check next week... no worries...

rene commented:
@eriknordmark , I have reworked this PR and addressed all my own comments :smile: 
I tested with a Raspberry PI 4 and the LED it's working as expected. I think you can fetch my branch https://github.com/rene/eve/tree/rpi-fix-led and push to yours (and of course, review my changes)...

eriknordmark commented:
> @eriknordmark , I have reworked this PR and addressed all my own comments smile I tested with a Raspberry PI 4 and the LED it's working as expected. I think you can fetch my branch https://github.com/rene/eve/tree/rpi-fix-led and push to yours (and of course, review my changes)...

@rene easier if you submit a PR from your branch and I can review and approve it.

rene commented:
@eriknordmark , here is the new PR: https://github.com/lf-edge/eve/pull/3067

eriknordmark commented:
Superceeded by #3067
